### PR TITLE
Feature/68

### PR DESCRIPTION
**Pull request by AAndersn**

### CODE REVIEW

**Comment by AAndersn** [comment]
Addresses issues 68 through 79

**Comment by AAndersn** [comment]
fix #56

**Comment by AAndersn** [comment]
fixes #37 as well

**Review by afisher1 — CHANGES_REQUESTED** [review]
I am requesting changes to address a couple bugs. cimhub_2023.py line 35:
current: if self.mRID is not None:
change: if self.mRID is not None and str(self.identifier) != self.mRID:

cimhub_2023.py lines 100-106 need to be outside of the if statement on line 96

File: cimgraph/data_profile/cimhub_2023/cimhub_2023.py [hunk 1]
```diff
@@ -0,0 +1,18692 @@
...
+        metadata = {
+            'type': 'Attribute',
+            'minOccurs': '1',
+            'maxOccurs': '1'
+        })
+
+    # Backwards support for objects created with mRID
+    def __post_init__(self) -> None:
+        if 'mRID' in self.__dataclass_fields__:
+            if self.mRID is not None:
```
Comment (afisher1) [inline cimgraph/data_profile/cimhub_2023/cimhub_2023.py:35]:
```

File: cimgraph/data_profile/cimhub_2023/cimhub_2023.py [hunk 2]
```diff
@@ -0,0 +1,18692 @@
...
+                self.__uuid__.uri_has_underscore = True
+            if uri.lower() != uri:
+                self.__uuid__.uri_is_capitalized = True
+            try:
+                self.identifier = UUID(uri.strip('_').lower())
+                invalid_mrid = False
+            except:
+                seed = seed + uri
+                _log.warning(f'URI {uri} not a valid UUID, generating new UUID')
+        if mRID is not None and str(self.identifier) != self.mRID:
```
Comment (afisher1) [inline cimgraph/data_profile/cimhub_2023/cimhub_2023.py:90]:
```suggestion
        if mRID is not None:
            if str(self.identifier) != mRID:
                # Handle inconsistent capitalization / underscores
                if mRID.strip('_') != mRID:
                    self.__uuid__.mrid_has_underscore = True
                    if uri is None:
                        self.__uuid__.uri_has_underscore = True
                if mRID.lower() != mRID:
                    self.__uuid__.mrid_is_capitalized = True
                    if uri is None:
                        self.__uuid__.uri_is_capitalized = True
                try:
                    self.identifier = UUID(mRID.strip('_').lower())
                    invalid_mrid = False
                except:
                    seed = seed + mRID
                    _log.warning(f'mRID {mRID} not a valid UUID, generating new UUID')
            else:
                invalid_mrid = False
            if 'mRID' in self.__dataclass_fields__:
                self.mRID = mRID
```

File: cimgraph/data_profile/cimhub_2023/cimhub_2023.py [hunk 3]
```diff
@@ -0,0 +1,18692 @@
...
+    identifier: Optional[UUID] = field(
+        default = None,
+        metadata = {
+            'type': 'Attribute',
+            'minOccurs': '1',
+            'maxOccurs': '1'
+        })
+
+    # Backwards support for objects created with mRID
+    def __post_init__(self) -> None:
```
Comment (afisher1) [inline cimgraph/data_profile/cimhub_2023/cimhub_2023.py:33]: As I see it there are a couple reasons to call self.uuid() in post_init.

1. self.identifier is None.
2. self.identifer doesn't match self.mRID.

Currently if the dataclass instance has self.mRID it calls self.uuid() no matter what. This call is completely unnecessary if a user specified identifier and mRID such that they match and the is the source of the bug I'm seeing where identifier if different from what I specified in the constructor.

**Review by afisher1 — CHANGES_REQUESTED** [review]

File: cimgraph/data_profile/cimhub_2023/cimhub_2023.py (hunk 3 above)
Comment (afisher1) [inline cimgraph/data_profile/cimhub_2023/cimhub_2023.py:33]:
```suggestion
    def __post_init__(self) -> None:
        if self.identifier is None:
            if 'mRID' in self.__dataclass_fields__:
                if self.mRID is not None:
                    self.uuid(mRID = self.mRID)
                else:
                    self.uuid()
            else:
                self.uuid()
        else:
            if 'mRID' in self.__dataclass_fields__:
                if str(self.identifier) != self.mRID:
                    if self.mRID is not None:
                        self.uuid(mRID = self.mRID)
                    else:
                        self.mRID = str(self.identifier)
```

File: cimgraph/data_profile/cimhub_2023/cimhub_2023.py (hunk 2 above)
Comment (afisher1) [reply cimgraph/data_profile/cimhub_2023/cimhub_2023.py:90]: The suggested change in here fixes an issue where the user/post_init calls uuid(mRID=) such that mRID = str(identifier) and invalid_mrid never gets set to False thus triggering another UUID generation and overwrites self.identifier.

File: cimgraph/data_profile/cimhub_2023/cimhub_2023.py [hunk 4]
```diff
@@ -0,0 +1,18708 @@
...
+                    self.uuid()
+            else:
+                self.uuid()
+        else:
+            if 'mRID' in self.__dataclass_fields__:
+                if str(self.identifier) != self.mRID:
+                    if self.mRID is not None:
+                        self.uuid(mRID=self.mRID)
+                    else:
+                        self.mRID = str(self.identifier)
```
Comment (afisher1) [inline cimgraph/data_profile/cimhub_2023/cimhub_2023.py:48]:
```suggestion
                        self.uuid(mRID=str(self.identifier))
```

File: cimgraph/data_profile/cimhub_2023/cimhub_2023.py [hunk 5]
```diff
@@ -0,0 +1,18708 @@
...
+        invalid_mrid = True
+        self.__uuid__ = self.__uuid_meta__()
+        # If URI is specified, try creating from UUID from URI
+        if uri is not None:
+            # Handle inconsistent capitalization / underscores
+            if uri.strip('_') != uri:
+                self.__uuid__.uri_has_underscore = True
+            if uri.lower() != uri:
+                self.__uuid__.uri_is_capitalized = True
+            try:
```
Comment (afisher1) [inline cimgraph/data_profile/cimhub_2023/cimhub_2023.py:96]:
```suggestion
            if str(self.identifier) != uri:
                try:
                    self.__uuid__.uuid = UUID(uri.strip('_').lower())
                    self.identifier = self.__uuid__.uuid
                    invalid_mrid = False
                except:
                    seed = seed + uri
                    _log.warning(f'URI {uri} not a valid UUID, generating new UUID')
            else:
                self.__uuid__.uuid = self.identifier
                invalid_mrid = False
```

File: cimgraph/data_profile/cimhub_2023/cimhub_2023.py [hunk 6]
```diff
@@ -0,0 +1,18708 @@
...
+                self.identifier = UUID(uri.strip('_').lower())
+                invalid_mrid = False
+            except:
+                seed = seed + uri
+                _log.warning(f'URI {uri} not a valid UUID, generating new UUID')
+        if mRID is not None:
+            if str(self.identifier) != mRID:
```
Comment (afisher1) [inline cimgraph/data_profile/cimhub_2023/cimhub_2023.py:103]:
```suggestion
            # Handle inconsistent capitalization / underscores
            if mRID.strip('_') != mRID:
                self.__uuid__.mrid_has_underscore = True
                if uri is None:
                    self.__uuid__.uri_has_underscore = True
            if mRID.lower() != mRID:
                self.__uuid__.mrid_is_capitalized = True
                if uri is None:
                    self.__uuid__.uri_is_capitalized = True
            if str(self.identifier) != mRID:
                try:
                    self.__uuid__.uuid = UUID(mRID.strip('_').lower())
                    self.identifier = self.__uuid__.uuid
                    invalid_mrid = False
                except:
                    seed = seed + mRID
                    _log.warning(f'mRID {mRID} not a valid UUID, generating new UUID')
            else:
                self.__uuid__.uuid = self.identifier
                invalid_mrid = False
            if 'mRID' in self.__dataclass_fields__:
                self.mRID = mRID
```

File: cimgraph/data_profile/cimhub_2023/cimhub_2023.py [hunk 7]
```diff
@@ -0,0 +1,18708 @@
...
+                except:
+                    seed = seed + mRID
+                    _log.warning(f'mRID {mRID} not a valid UUID, generating new UUID')
+            else:
+                invalid_mrid = False
+            if 'mRID' in self.__dataclass_fields__:
+                self.mRID = mRID
+        # Otherwise, build UUID using unique name as a seed
+        if invalid_mrid:
+            if name is not None:
```
Comment (afisher1) [inline cimgraph/data_profile/cimhub_2023/cimhub_2023.py:125]:
```suggestion
            if not self.identifier:
                if name is not None:
                    seed = seed + f'{self.__class__.__name__}:{name}'
                    randomGenerator = Random(seed)
                    self.__uuid__.uuid = UUID(int=randomGenerator.getrandbits(128), version=4)
                    self.name = name
                else:
                    self.__uuid__.uuid = uuid4()
                self.identifier = self.__uuid__.uuid
                # Write mRID string for backwards compatibility
                if 'mRID' in self.__dataclass_fields__:
                    if mRID is not None:
                        self.mRID = mRID
                    else:
                        self.mRID = str(self.identifier)
            else:
                self.__uuid__.uuid = self.identifier 
                # Write mRID string for backwards compatibility
                if 'mRID' in self.__dataclass_fields__:
                    if mRID is not None:
                        self.mRID = mRID
                    else:
                        self.mRID = str(self.identifier)
```

**Review by afisher1 — CHANGES_REQUESTED** [review]

**Comment by AAndersn** [comment]
fixes #82

**Comment by AAndersn** [comment]
fixes #79

**Review by afisher1 — APPROVED** [review]

**Pull request merged**